swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/IPCClient.swift

```diff
@@ -97,39 +97,130 @@ enum IPCClient {
     try await sendMessageWithoutResponse(session: session, message: message)
   }
 
-  static func clearLogs(session: NETunnelProviderSession) async throws {
-    try await sendMessageWithoutResponse(session: session, message: ProviderMessage.clearLogs)
+  // MARK: - Log operations with utun cleanup
+
+  #if os(macOS)
+    /// Wraps an IPC call with wake-up and cleanup logic when VPN is not connected.
+    ///
+    /// On macOS, IPC calls can wake the network extension, which creates a utun device.
+    /// If not properly cleaned up (via stop), these devices accumulate.
+    @MainActor
+    private static func wrapIPCCallIfNeeded<T: Sendable>(
+      session: NETunnelProviderSession,
+      configuration: Configuration,
+      operation: @MainActor () async throws -> T
+    ) async throws -> T {
+      if session.status == .connected {
+        // Extension already running, utun already created, just execute
+        return try await operation()
+      }
+
+      // Extension needs waking, must cleanup afterwards
+      try dryStartStopCycle(session: session, configuration: configuration)
```

Comment:
Doesn't this already start *and* stop the tunnel? I.e. I don't think the stop below is needed?

The completionHandler in the startTunnel on the PacketTunnelProvider side returns a dummy error which is the same as calling stop I believe.

Reply:
On this note, if that's the case, we start-stop twice:

dryStartStopCycle and then again implicitly via the IPC call, and then a final stop.

@klochowicz Do you mind if I push a commit to this branch? I think that'll be faster than providing the feedback here.